### PR TITLE
[SYCLomatic] Add dpct option check architecture

### PR DESCRIPTION
**Pull request by ziranzha**

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/DpctOptions.cpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
+#include "clang/DPCT/DpctOptions.h"
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DpctOptions.cpp:9]: Please add license header for this new file.

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 2]
```diff
@@ -1,564 +1,979 @@
-#ifndef DPCT_OPT_TYPE
-#define DPCT_OPT_TYPE(...)
+#ifdef DPCT_OPTIONS_VAR
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc:24]: Please add license header for this file.

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 3]
```diff
@@ -1,564 +1,979 @@
...
+
+#if DPCT_OPTIONS_IN_CLANG_DPCT
+
+#define DPCT_OPTION(TEMPLATE, TYPE, NAME, ...)                                 \
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc:28]: Please add comment for the DPCT_OPTION macro on functionality

File: clang/lib/DPCT/DpctOptions.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  DependedOptions |= Dependencies;
+}
+
+void DpctOptionBase::init() {
+  for (auto Option : DpctOptionList) {
+    if (!Option)
+      continue;
+    Option->setExclusiveByAction();
+  }
+  setDependency(DpctOptionNameKind::OPT_ProcessAll,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DpctOptions.cpp:174]: do we have more dependency to add?

File: clang/lib/DPCT/DpctOptions.cpp [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+inline void DpctOptionBase::setExclusive(DpctOptionNameKind OptionA,
+                                         DpctOptionNameKind OptionB) {
+  DpctOptionBase *OptA = getOption(OptionA), *OptB = getOption(OptionB);
+  OptA->Exclusive |= OptB->FlagBit;
+  OptB->Exclusive |= OptA->FlagBit;
+}
+
+inline void DpctOptionBase::setDependency(DpctOptionNameKind DependentOption,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DpctOptions.cpp:104]: Please add comment when need call this API.

File: clang/include/clang/DPCT/DpctOptions.h [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  DpctOptionClass Class;
+  BitsType FlagBit;
+  BitsType Exclusive;
+  BitsType Dependencies;
+
+  static std::vector<DpctOptionBase *> getOptions(BitsType Flag);
+
+  static void setDependency(DpctOptionNameKind DependentOption,
+                            DpctOptionNameKind DependedOption);
+  static void setExclusive(DpctOptionNameKind OptionA,
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DpctOptions.h:66]: have not found use of setExclusive(), please comment when to use this API.

File: clang/lib/DPCT/DpctOptions.cpp (hunk 4 above)
Comment (ziranzha) [reply clang/lib/DPCT/DpctOptions.cpp:174]: Currently, there is no more.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**